src/v/kafka/server/usage_aggregator.cc

```diff
@@ -0,0 +1,459 @@
+    }
+    if (!period && !windows && !data) {
+        /// Data didn't exist
+        return std::nullopt;
+    } else if (!period || !windows || !data) {
+        vlog(
+          klog.error,
+          "Inconsistent usage_manager on disk state detected, failed to "
+          "recover state");
+        return std::nullopt;
+    }
+    return persisted_state{
+      .configured_period = serde::from_iobuf<std::chrono::seconds>(
+        std::move(*period)),
+      .configured_windows = serde::from_iobuf<size_t>(std::move(*windows)),
+      .current_state = serde::from_iobuf<fragmented_vector<usage_window>>(
+        std::move(*data))};
+}
+
+static ss::future<> clear_persisted_state(storage::kvstore& kvstore) {
+    using kv_ks = storage::kvstore::key_space;
+    try {
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(period_key));
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(max_duration_key));
+        co_await kvstore.remove(kv_ks::usage, key_to_bytes(buckets_key));
+    } catch (const std::exception& ex) {
+        vlog(klog.debug, "Ignoring exception from storage layer: {}", ex);
+    }
+}
+
+usage usage::operator+(const usage& other) const {
+    return usage{
+      .bytes_sent = bytes_sent + other.bytes_sent,
+      .bytes_received = bytes_received + other.bytes_received};
+}
+
+usage& usage::operator+=(const usage& other) {
+    if (this != &other) {
+        bytes_sent += other.bytes_sent;
+        bytes_received += other.bytes_received;
+        bytes_cloud_storage = other.bytes_cloud_storage;
+    }
+    return *this;
+}
+
+template<typename time_point>
+auto epoch_time_secs(time_point now) {
+    return std::chrono::duration_cast<std::chrono::seconds>(
+             now.time_since_epoch())
+      .count();
+}
+
+static ss::lowres_system_clock::time_point round_to_interval(
+  std::chrono::seconds usage_window_width_interval,
+  ss::lowres_system_clock::time_point t) {
+    /// Downstream systems are particularly sensitive to minor issues with
+    /// timestamps not triggering on the configured interval (hours, minutes,
+    /// seconds, etc), this method rounds t to the nearest interval and logs an
+    /// error if this cannot be done within some threshold.
+    static const auto err_threshold = 5s;
+    const auto interval = usage_window_width_interval;
+    const auto cur_interval_start = t - (t.time_since_epoch() % interval);
+    const auto next_interval_start = cur_interval_start + interval;
+    if (t - cur_interval_start <= err_threshold) {
+        return {cur_interval_start};
+    } else if (next_interval_start - t <= err_threshold) {
+        return {next_interval_start};
+    }
+    vlog(
+      klog.error,
+      "usage has detected a timestamp '{}' that exceeds the preconfigured"
+      "threshold of 5s meaning a clock has fired later then expected, "
+      "this is unexpected behavior and should be investigated.",
+      t.time_since_epoch().count());
+    return t;
```

Comment:
@alenkacz how do you think we should behave in the case there is some type of internal error which prevented the window from closing too early or late , exceeding a hardcoded value of +- 5s? We can return the rounded timestamp or the actual time recorded

Reply:
I think we already discussed, you preferred option 2. I can work with both as long as we don't roll over the drift (in case of option 2). Btw. in terms of further processing both options are the same to me, we always have to assign the usage to whole hour so either redpanda or the processing pipeline will have to lie.

### Option 1
you close at the end of hour and start new one at the aligned timestamp (so you'll be lying - this could be lying potentially by hours as well)

window started 10:00
broker came up at 11:15
broker closes window - so it will be 10:00 to 11:00 (it will not tell truth and assign the 15m usage to previous hour potentially)
opens next window at 11:00
closes this window in shorter time in 12:00

### Option 2
in rare cases there could be misaligned windows as long as it's rare and we don't roll over this error further. By the way even with this solution you will be more truthful in the reporting but we'll be lying exactly in the same way as Option 1 - we can always only account the usage to one whole hour so we will anyway always assign the whole usage to one aligned window.

window started 10:00
broker came up at 11:15
broker closes window - so it will be 10:00 to 11:15
opens next window at 11:15
closes this window in shorter time in 12:00 (!) this is important

Reply:
I think it makes more sense to go with option 1, since in all other paths you'd never receive an unaligned window

Reply:
let me know which option is implemented

Reply:
I'm now having second thoughts. We have things that we want to ensure, 
1. That the timer fires on the top of the hour (irrespective of last timestamp, so we avoid skew)
2. That timestamps are aligned and make sense (end - begin should be one hour and next begin should == prev_end so on and so forth)

In the event we have drifted so far there is an issue, what can we do?
1. Force an alignment, this means we would fire the timer at the right time, but then we could end up in a situation like this:
- Timer fires at 12:30pm
- We move this value to 1:00pm
- Timer then fires at 1:00pm
- Begin and end timestamps are the same for this window
Or 
- Timer fires at 12:30pm
- We move this value to 12:00pm
- Now begin and end timestamps of previous window are identical
2. Force an alignment but allow timer to drift. This would mean the reported values in the buckets aren't reflecting the real world situation

I don't think either situation is acceptable, I think in these rare cases it makes more sense to just leave the timestamp as it is. @dotnwat

Reply:
Here is a proposal that I think covers all the cases. WDYT?

---

* Set a timer to fire at the top of every hour
* When the timer fires
    * If validation[0] passes, then
        * Finalize the usage for the current window
        * Window timestamp reported is the target top-of-hour
    * Else
        * Skip window, report errors, log errors, alert, etc...
        * Reset system to next valid window

[0] validation: of course a timer will never fire exactly on time. in particular, timers only guarantee that they fire at a time >= the target time. in practice we need to define what is acceptable and deal with scenarios that fall outside of a band of ranges. some examples of orders of magnitude error might be:

    a. [-1s, +1s] - covers basic scheduling variability, low resolution timers, some reactor stalls, etc...
    b. [-2min, +2min] - extreme anomaly. deadlocks. infinite loops. the system is probably already unusable.
    c. > |2min| - whatever is worse than extreme

I think that for (a) and (b) we can round to the top of the hour and report as normal. For (c) we should probably just drop the window and sure an alert is noticed by an administrator.

Implementation notes

1. when rounding preserve the property that usage data accounted for in exactly 1 window (or 0 if handling exceptional cases).
2. use the system clock for scheduling the next timer, and a monotonic clock for determining errors to account for cases in which you accidentally caught a big NTP update. or if this is not a real concern, you could treat this case as the rare exceptional case and drop the window.
3. enforce the rule that the collection interval is always greater than the acceptable error range

Reply:
I guess the options are:
1. Unaligned window with some data
2. Dropped window altogether, so no data

Reply:
I'd characterize it more like:

1. Always aligned window
2. Extremely rare dropped window (how would you characterize extremely rare?)

Even in case (2) you could flip a coin and put the usage data in the current or next window. I don't have any opinion about that and I think it is more related to how rare the case is and to what extent there is value in only having aligned windows.

Reply:
with Rob on slack we agreed on option 2 from my original proposal - so having very rarely unaligned windows that will close at the end of an hour. 

Internally we use start time as a way of assigning bucket to an hour and we strip minutes and seconds so it should work anyway (this way we'll assign the usage to the original hour even when the bucket ended e.g. 11:30).

This way if this is not acceptable, we can start droping such windows in the metering pipeline and make decision on our own. So I prefer that to redpanda droping the windows by itself at the source.